fvm/environment/programs.go

```diff
@@ -215,45 +211,84 @@ func (programs *Programs) getOrLoadAddressProgram(
 		return nil, fmt.Errorf("get program failed: %w", freezeError)
 	}
 
-	// reading program from cache
-	program, programState, has := programs.txnState.GetProgram(address)
-	if has {
+	loader := newProgramLoader(load, programs.dependencyStack)
+	program, err := programs.txnState.GetOrComputeProgram(
+		programs.txnState,
+		address,
+		loader,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error getting program: %w", err)
+	}
+
+	programs.dependencyStack.addDependencies(program.Dependencies)
```

Comment:
I feel we could move this into the `if loader.Called()`. 

If loader was not called, it means the cache has the program already, so we don't need to add it again, right?